### ISSUE TITLE
[BUG] Intellij shows error "Missing required property $ref" in AsyncAPI 3.0.0 document

### ISSUE THREAD

**Issue by timonback** [asyncapi/jasyncapi-idea-plugin#49]
### Describe the bug.

The provided specs are valid AsyncAPI to the best of our knowledge. The `asyncapi/parser` + AsyncAPI Studio shows these files as valid.

However, the Intellij plugin shows errors

### Expected behavior

No errors in the AsyncAPI plugin.

Similar or identical validation to studio.asyncapi.com

### Screenshots

-

### How to Reproduce

Open one of the example files in Springwolf, for example:
- https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-sqs-example/src/test/resources/asyncapi.json
- https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json

### 🥦 Browser

None

### 👀 Have you checked for similar open issues?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md)

### Are you willing to work on this issue ?

No, someone else can work on it

**Comment by Pakisan** [comment]
> ### Describe the bug.
> The provided specs are valid AsyncAPI to the best of our knowledge. The `asyncapi/parser` + AsyncAPI Studio shows these files as valid.
> 
> However, the Intellij plugin shows errors
> 
> ### Expected behavior
> No errors in the AsyncAPI plugin.
> 
> Similar or identical validation to studio.asyncapi.com
> 
> ### Screenshots
> 
> ### How to Reproduce
> Open one of the example files in Springwolf, for example:
> 
> * https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-sqs-example/src/test/resources/asyncapi.json
> * https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-amqp-example/src/test/resources/asyncapi.json
> 
> ### 🥦 Browser
> None
> 
> ### 👀 Have you checked for similar open issues?
> * [x]  I checked and didn't find similar issue
> 
> ### 🏢 Have you read the Contributing Guidelines?
> * [x]  I have read the [Contributing Guidelines](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md)
> 
> ### Are you willing to work on this issue ?
> No, someone else can work on it

Will check it today. It may be bug in editor or validator

**Comment by theschles** [comment]
@Pakisan confirmed that the asyncapi files to which @timonback provided links both pass validation:

```
springwolf-core/springwolf-examples/springwolf-sqs-example/src/test/resources on  master via ⬡ v8.6 via ☕️ v17.0.5
➜ asyncapi --version
@asyncapi/cli/1.6.1 darwin-arm64 node-v20.10.0

springwolf-core/springwolf-examples/springwolf-sqs-example/src/test/resources on  master via ⬡ v8.6 via ☕️ v17.0.5
➜ asyncapi validate asyncapi.json

File asyncapi.json is valid! File asyncapi.json and referenced documents don't have governance issues.

springwolf-core/springwolf-examples/springwolf-sqs-example/src/test/resources on  master via ⬡ v8.6 via ☕️ v17.0.5
➜ cd ../../../../springwolf-amqp-example/src/test/resources

springwolf-core/springwolf-examples/springwolf-amqp-example/src/test/resources on  master via ⬡ v8.6 via ☕️ v17.0.5
➜ asyncapi validate asyncapi.json

File asyncapi.json is valid! File asyncapi.json and referenced documents don't have governance issues.
```

**Comment by theschles** [comment]
Here are some of the problems seemingly reported by the plugin:

<img width="848" alt="image" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/901430/e50ccb67-2329-4a40-82c2-190ff43c1d95">

...and...

<img width="851" alt="image" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/901430/aa4f4eec-69f8-4382-9dc0-336221f7c202">

**Comment by Pakisan** [comment]
@theschles thanks)

I have built new version, at night, with updated schema. Will check on todays morning

**Comment by Pakisan** [comment]
@theschles @timonback not easy one problem. Awesome mix of bugs in IDEA related to `anyOf`, `oneOf` and issues in our Json Schemas.

Need more time to dig into `@asyncapi/cli` and figure out which schemas they are using for validation

**Comment by Pakisan** [comment]
ok. looks like problem is in `payload` validation:

it's ok
```json
"io.github.springwolf.examples.amqp.dtos.AnotherPayloadDto": {
        "headers": {
          "$ref": "#/components/schemas/HeadersNotDocumented"
        },
        "payload": {
          "schemaFormat": "application/vnd.aai.asyncapi+json;version=3.0.0"
        },
        "name": "io.github.springwolf.examples.amqp.dtos.AnotherPayloadDto",
        "title": "AnotherPayloadDto",
        "description": "Another payload model",
        "bindings": {
          "amqp": {
            "bindingVersion": "0.3.0"
          }
        }
      },
```

it's not ok
```json
"io.github.springwolf.examples.amqp.dtos.ExamplePayloadDto": {
        "headers": {
          "$ref": "#/components/schemas/HeadersNotDocumented"
        },
        "payload": {
          "schemaFormat": "application/vnd.aai.asyncapi+json;version=3.0.0",
          "schema": {
            "$ref": "#/components/schemas/io.github.springwolf.examples.amqp.dtos.ExamplePayloadDto"
          }
        },
        "name": "io.github.springwolf.examples.amqp.dtos.ExamplePayloadDto",
        "title": "ExamplePayloadDto",
        "bindings": {
          "amqp": {
            "bindingVersion": "0.3.0"
          }
        }
      }
```

**Comment by Pakisan** [comment]
ampq example was fixed

sqs on a way

**Comment by Pakisan** [comment]
@timonback @theschles both examples were fixed.

I'll push updated schemas to release branch. It'll help you to re-check fix after installation of  local built plugin

**Comment by Pakisan** [comment]
@timonback @theschles build to test
[jasyncapi-idea-plugin-2.5.0+jre17.zip](https://github.com/asyncapi/jasyncapi-idea-plugin/files/14615243/jasyncapi-idea-plugin-2.5.0%2Bjre17.zip)

**Comment by timonback** [comment]
Thanks @Pakisan , the `$ref` errors are mostly gone - or at least not errors anymore

I do see the following warnings, not sure if they are expected with the plugin. I am using again https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-sqs-example/src/test/resources/asyncapi.json as it is the smallest file:

![image](https://github.com/asyncapi/jasyncapi-idea-plugin/assets/7568775/37f0f219-fa7f-4379-83a9-2e9b48b7a8a0)
and
![image](https://github.com/asyncapi/jasyncapi-idea-plugin/assets/7568775/6c3ee295-b1a9-49ee-8659-b69ab2489519)

My Intellij didn't know about AsyncApi3 json-schema yet, so I had to manually define a file `JSON Schema Mapping` using this schema: https://www.asyncapi.com/definitions/3.0.0.json
Otherwise, the 2.X.0 version was used.

**Comment by Pakisan** [comment]
> Thanks @Pakisan , the `$ref` errors are mostly gone - or at least not errors anymore
> 
> I do see the following warnings, not sure if they are expected with the plugin. I am using again https://github.com/springwolf/springwolf-core/blob/master/springwolf-examples/springwolf-sqs-example/src/test/resources/asyncapi.json as it is the smallest file:
> 
> ![image](https://private-user-images.githubusercontent.com/7568775/313202042-37f0f219-fa7f-4379-83a9-2e9b48b7a8a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MTAwOTgsIm5iZiI6MTcxMDUwOTc5OCwicGF0aCI6Ii83NTY4Nzc1LzMxMzIwMjA0Mi0zN2YwZjIxOS1mYTdmLTQzNzktODNhOS0yZTliNDhiN2E4YTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTVUMTMzNjM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDcwZmIyMWI5ZDE5NTFiMDBkY2I3ODVhOGY2OGI1MzljNDZjNjlkMmI3ZmQ1Zjc1MDM1YWZmYzY3MTI4OTA0NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.uAJp_sky2Ih6SRCUrAnG18qeANU0G4r-UrLzYV1FK4k) and ![image](https://private-user-images.githubusercontent.com/7568775/313202237-6c3ee295-b1a9-49ee-8659-b69ab2489519.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA1MTAwOTgsIm5iZiI6MTcxMDUwOTc5OCwicGF0aCI6Ii83NTY4Nzc1LzMxMzIwMjIzNy02YzNlZTI5NS1iMWE5LTQ5ZWUtODY1OS1iNjlhYjI0ODk1MTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTVUMTMzNjM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjBlMzVkMzA0ZmI0Y2FhMGE5MzI3OTcwNzE0Nzg1NzAzNmFmZTc3ZDZjMGU5ZGZlZTAzZTFhYzI4ODA5ODRlOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.F6ZwYYSoDAputNr90mKVWd4qmydCC6rpShg2B3u_QLY)
> 
> My Intellij didn't know about AsyncApi3 json-schema yet, so I had to manually define a file `JSON Schema Mapping` using this schema: https://www.asyncapi.com/definitions/3.0.0.json Otherwise, the 2.X.0 version was used.

No, no, no, you don't need to define external json schema. It's already bundled in plugin

Download archive and install update manually: plugins -> gear wheel icon -> install from disk

<img width="1512" alt="Screenshot 2024-03-15 at 17 40 14" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/3388414/bf7c5e04-f844-4820-b563-0f4d26c8705b">

**Comment by timonback** [comment]
Just checked it on a second machine as well (version 2.5.0+jre17):
- [x] No extra mapping schema necessary
- [ ] Still it (amqp in this case) shows problems

For the schema (status bar), it states: `JSON: all.schema-store.json`

![image](https://github.com/asyncapi/jasyncapi-idea-plugin/assets/7568775/2360a827-eeaa-431f-b1af-1dcb18f97663)

**Comment by Pakisan** [comment]
@theschles please, delete mapping with any schemas, because you are using not fixed schema from https://www.asyncapi.com/schema-store/all.schema-store.json

<img width="1094" alt="Screenshot 2024-03-15 at 21 13 15" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/3388414/412d3d89-909d-41fb-81bb-782a163fe08d">
<img width="1512" alt="Screenshot 2024-03-15 at 21 13 47" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/3388414/0772fe55-f982-4356-8f46-bea063267b8b">
<img width="1512" alt="Screenshot 2024-03-15 at 21 14 00" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/3388414/56cfa106-2b10-4ca2-87f7-f1dca541ad90">

**Comment by timonback** [comment]
Thanks @Pakisan , that does the trick.

I don't think - although I might be wrong - that I have touched the default Intellij setting.
I guess it is a requirement for the AsyncApi idea plugin to disable to the "Remote JSON Schemas" > "Allow download JSON Schemas from remote sources"?

Anyway, the issue is resolved from my point of view.
Many thanks again 👍

**Comment by Pakisan** [comment]
> Thanks @Pakisan , that does the trick.
> 
> I don't think - although I might be wrong - that I have touched the default Intellij setting. I guess it is a requirement for the AsyncApi idea plugin to disable to the "Remote JSON Schemas" > "Allow download JSON Schemas from remote sources"?
> 
> Anyway, the issue is resolved from my point of view. Many thanks again 👍

> I don't think - although I might be wrong - that I have touched the default Intellij setting.
> I guess it is a requirement for the AsyncApi idea plugin to disable to the "Remote JSON Schemas" > "Allow download JSON Schemas from remote sources"?

This trick will not be required after merging of this MR - https://github.com/asyncapi/spec-json-schemas/pull/495. Until that my fix is not available from our side (https://www.asyncapi.com/schema-store/all.schema-store.json)

**Comment by theschles** [comment]
> @theschles please, delete mapping with any schemas, because you are using not fixed schema from https://www.asyncapi.com/schema-store/all.schema-store.json
> <img alt="Screenshot 2024-03-15 at 21 13 15" width="1094" src="https://private-user-images.githubusercontent.com/3388414/313279028-412d3d89-909d-41fb-81bb-782a163fe08d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTgzMDUsIm5iZiI6MTcxMDY5ODAwNSwicGF0aCI6Ii8zMzg4NDE0LzMxMzI3OTAyOC00MTJkM2Q4OS05MDlkLTQxZmItODFiYi03ODJhMTYzZmUwOGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTdUMTc1MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWVjYjQ4YTI2MWFmOTVmNTk1ZGI4NzNiNDUxNzE0OGFjY2E2ODFhOTUzOWZlM2NkNDc3Yjc5NzNlY2QxNjZjMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8Ozy8mHws2jPppzW7hggdgsyKB8M5gtEFRBC6JImbug"> <img alt="Screenshot 2024-03-15 at 21 13 47" width="1512" src="https://private-user-images.githubusercontent.com/3388414/313279036-0772fe55-f982-4356-8f46-bea063267b8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTgzMDUsIm5iZiI6MTcxMDY5ODAwNSwicGF0aCI6Ii8zMzg4NDE0LzMxMzI3OTAzNi0wNzcyZmU1NS1mOTgyLTQzNTYtOGY0Ni1iZWEwNjMyNjdiOGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTdUMTc1MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzRmYmQwODJjZGZlNTZiNzY2ZDIxNDVlZGY0MWU4YmY3MWIyMGU0ZjlmMmM3YjA2NGFlYWQ4OGFjYTIzZDJkOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3cX90IxyvgzcyXPh0bZRkRWDF6FkTirEoyzywEJM-Xs"> <img alt="Screenshot 2024-03-15 at 21 14 00" width="1512" src="https://private-user-images.githubusercontent.com/3388414/313279059-56cfa106-2b10-4ca2-87f7-f1dca541ad90.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2OTgzMDUsIm5iZiI6MTcxMDY5ODAwNSwicGF0aCI6Ii8zMzg4NDE0LzMxMzI3OTA1OS01NmNmYTEwNi0yYjEwLTRjYTItODdmNy1mMWRjYTU0MWFkOTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTdUMTc1MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yjg3ZWE4NTM1ZWE3MWIyYjczZmYwOTA3ODI1MmFhY2JmOGM2Y2M1N2I3NmEyZThmNzFhYmU1NDZjOTJmZDYyNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8FwcqLryHZu-kirW1S8l3j9iuopk7P5CLBZ7QxHhva8">

Hi @Pakisan @timonback -- sorry, have been heads down at work all week due to schedule pressure. Now seeing the updates to this convo.

I didn't have any specified JSON Schema Mappings at all:

<img width="980" alt="image" src="https://github.com/asyncapi/jasyncapi-idea-plugin/assets/901430/c363270c-0942-4d06-bd39-5b6dab95b1dc">

**Comment by theschles** [comment]
> > Thanks @Pakisan , that does the trick.
> > I don't think - although I might be wrong - that I have touched the default Intellij setting. I guess it is a requirement for the AsyncApi idea plugin to disable to the "Remote JSON Schemas" > "Allow download JSON Schemas from remote sources"?
> > Anyway, the issue is resolved from my point of view. Many thanks again 👍
> 
> Great to hear, that fix is working like expected. Time to release it)
> 
> > I don't think - although I might be wrong - that I have touched the default Intellij setting.
> > I guess it is a requirement for the AsyncApi idea plugin to disable to the "Remote JSON Schemas" > "Allow download JSON Schemas from remote sources"?
> 
> This trick will not be required after merging of this MR - [asyncapi/spec-json-schemas#495](https://github.com/asyncapi/spec-json-schemas/pull/495). Until that my fix is not available from our side (https://www.asyncapi.com/schema-store/all.schema-store.json)

[Hopefully we'll get the remaining approval(s) tomorrow](https://github.com/asyncapi/spec-json-schemas/pull/495#issuecomment-1999871395)

> Please @Pakisan hold on so I can review this PR on Monday. I really need to understand the implication of this change (probably will be OK, but need to understand as I worked on that particular part of the JSON Schema recently).

**Comment by theschles** [comment]
Just to keep @Pakisan and @timonback in the loop - [the reviewer asked for some changes](https://github.com/asyncapi/spec-json-schemas/pull/495#pullrequestreview-1938924591)

**Comment by theschles** [comment]
@Pakisan  @timonback was this correctly resolved by the change in [`asyncapi/spec-json-schemas#495`?](https://github.com/asyncapi/spec-json-schemas/pull/495) ?

**Comment by timonback** [comment]
Thank you @Pakisan and @theschles , the issues are resolved!

Looking forward to the official release.
As mentioned before, it might be good to explicitly documented that users should disable the intellij schema store.

**Comment by Pakisan** [comment]
> Thank you @Pakisan and @theschles , the issues are resolved!
> 
> 
> 
> Looking forward to the official release.
> 
> As mentioned before, it might be good to explicitly documented that users should disable the intellij schema store.

You can enable now IntelliJ schema store. I think that updated schema was published after merge

I hope that release with this fix and 2024.1 will be approved by JetBrains on Monday

**Comment by Pakisan** [comment]
@timonback new version now available for downloading
